Guidelines: translations (#153)
codewars/docs: content/recipes/authoring/translation.md

# Creating a Translation

This article contains a set of guidelines, which can be used by translators to create good translations to existing kata. They were collected to help ensure that translations are of sufficient quality and users' experience will be as good as possible.

The guidelines should be used by translators and reviewers to verify whether translations about to be introduced into Codewars are of sufficient quality. Conformance with them should be verified before the translation is published and before it's approved.

Failure to comply with the below guidelines is considered an issue and should be reported as such. In case of severe violations, the affected translation can be rejected or even deleted.

## General translation guidelines

- **Respect the [General Content Guidelines](/recipes/authoring/general/).**

> - **Translations should improve the overall quality of the kata.** The sub-optimal quality of existing, approved translations should not be used as an excuse for authoring a low-quality translation. In particular:

Reviewer: Great point!

Author: Thx

> - The description can be clarified if necessary.

Reviewer 2: ```suggestion
  - The description can be clarified if necessary.
```

  - The test suite may employ a different structure from that of other translations. **The [sample](/recipes/authoring/kata-snippets/sample-tests/) and [submission](/recipes/authoring/kata-snippets/full-tests/) tests (including [random tests](/recipes/authoring/kata-snippets/full-tests/#random-tests)) should conform to Codewars quality guidelines.**
  - If serious issues with other language versions are uncovered in the course of authoring a translation, they should be fixed or raised as appropriate issues in the kata discourse.

> ## Conformance to overall kata idea

Reviewer: Perhaps "conformity" is more typical than "conformance"?

Author: ```suggestion
## Conformity to overall kata idea
```

Well, it's not me who speaks English here, so I believe you know better :)

Translation should not substantially change overall idea of the kata. It should be as consistent as possible with the original version in terms of difficulty, required performance, and overall composition. In particular:

- **Respect the performance requirements of the kata.** If, for example, it's agreed that kata should accept only (sub)linear solutions, make sure that translated tests reject solutions of higher computational complexity. Languages and runtimes differ in terms of performance, so increase/decrease amount of test cases and input ranges when necessary.
- If possible, your translation should hold to **original author's idea**. In case of doubts, you can always ask them for opinion, or use initial language version as a baseline. However, especially for older kata, it may happen that initial version (and other older versions) are outdated, and concept of the kata evolved. You can ask kata contributors or other users for opinion and ideas how to proceed.

> - **Avoid translations to languages which cannot support the idea of the kata.** For example, there's no use in translating a big integer arithmetic kata to Python, which supports arbitrary precision integers natively. Do not translate kata into languages where difficulty would signifficantly differ: from high level languages to NASM or BF, or into lnaguages where kata can be solved with easy to use, advanced features of standard library (or other available libraries).

Reviewer 3: ```suggestion
- **Avoid translations to languages which cannot support the idea of the kata.** For example, there's no use in translating a big integer arithmetic kata to Python, which supports arbitrary precision integers natively. Do not translate kata into languages where the difficulty would significantly differ: from high level languages to NASM or BF, or into languages where the kata can be solved with easy to use, advanced features of the standard library (or other available libraries).
```

## Description

- **Respect guidelines for [Writing a Kata Description](/recipes/authoring/kata-snippets/description/).**

> - **Use [Codewars Markdown Extensions](/references/markdown/extensions/#sequential-code-blocks)** to organize language specific parts of description, if present. Add sections specific to your language if necessary.

Reviewer 3: ```suggestion
- **Use [Codewars Markdown Extensions](/references/markdown/extensions/#sequential-code-blocks)** to organize language specific parts of the description, if present. Add sections specific to your language if necessary.
```

> - If possible, **avoid displaying many language variants to the user**. Use conditional paragraphs to display information revelant to the currently viewed language.

Reviewer 4: ```suggestion
- If possible, **avoid displaying many language variants to the user**. Use [conditional paragraphs](references/markdown/extensions/#conditional-rendering) to display information revelant to the currently viewed language.
```

> - If the language version can pass tests successfuly only when run with some specific language runtime, it should be specified in the description.

Reviewer 4: I don't get this sentence => ?

You mean time constraints that are different from the timeout itself?

Author: ```suggestion
- If a translation can pass tests successfully only when run with some specific language runtime, it should be specified in the description.
```

There used to be cases when some kata happened (was designed to?) pass, for example, with C++ 17, but not with C++ 14. Or C# Mono runtime vs C# 8. For some reason, sometimes these language versions were still available (and sometimes preselected) in the trainer, but it was not possible to pass when they were used.

## Code components

> - **Respect the [General Coding Guidelines](/recipes/authoring/kata-snippets/coding-general/), any conventions imposed by your [language](/languages/) and guidelines specific to each component of the kata**, e.g. the [preloaded](/recipes/authoring/kata-snippets/preloaded/) section, [reference solution](/recipes/authoring/kata-snippets/reference-solution/), [sample tests](/recipes/authoring/kata-snippets/sample-tests/) and [submission tests](/recipes/authoring/kata-snippets/full-tests/).

Reviewer 4: ```suggestion
- **Respect the [General Coding Guidelines](/recipes/authoring/kata-snippets/coding-general/), conventions imposed by your [language](/languages/) and guidelines specific to each component of the kata**, e.g. the [preloaded](/recipes/authoring/kata-snippets/preloaded/) section, [reference solution](/recipes/authoring/kata-snippets/reference-solution/), [sample tests](/recipes/authoring/kata-snippets/sample-tests/) and [submission tests](/recipes/authoring/kata-snippets/full-tests/).
```

Loads of links in this one. That makes it very cumbersome to read. Maybe one single link toward "Kata elements/overwiew"?

Author: You mean difficult to read in markdown, or on rendered page?

Reviewer 4: rendered

Author: This part got reorganized, let's see how it turns out.

> - If possible, **sample tests and fixed tests should remain consistent between translations.** It may not always be feasible due to inherent differences in programming languages, but doing so can save follow-up questions on the language a user is attempting the kata in when a question on failing tests is raised in the kata discourse.

Reviewer 4: I'd reword this one the other way around:

> sample and fixed tests should be the exact same than...
> ...excpet in some rare cases where...

Author: ```suggestion
- **Sample tests and fixed tests should remain consistent between translations**, except in some rare cases when it's not feasible due to inherent differences in programming languages. It can save follow-up questions on the language a user is attempting the kata in when a question on failing tests is raised in the kata discourse.
```

> - Sometimes, description is not the best place to put some language specific information. For example, translator wants to add some information on arguments, data formats, memory management, etc. and keep the description language-agnostic at the same time. In such cases, it's allowed to put such information as an explanatory comment in "solution setup" snippet. 

Reviewer 3: ```suggestion
- Sometimes, the description is not the best place to put some language specific information. For example, the translator may wish to add some information on arguments, data formats, memory management, etc. and keep the description language-agnostic at the same time. In such cases, it's allowed to put such information as an explanatory comments in the solution setup.
```

## Review process

> - **Translation must be reviewed before it's approved.** If reviewer is not familiar with the language of the translation, they should ask other users for help.

Reviewer 3: ```suggestion
- **The translation must be reviewed before it's approved.** If the reviewer is not familiar with the language of the translation, they should ask other users for help.
```

> - **The translation must be reviewed before it's approved.** If the reviewer is not familiar with the language of the translation, they should ask other users for help.

Reviewer 4: ```suggestion
- **Translations must be reviewed before they are approved.** If the reviewer is not familiar with the language of the translation, they should ask other users for help.
```

Reviewer 4: ok, I begin to get a better idea about your approch when writing this part. Removing plural, so. Forget about this one.

> - **Any issues found during review should be posted** as translation comments.

Reviewer 4: shouldn't that be "any issue"? (singular)

I'm not entirely sure about the meaning of "issue" in this sentence. If that's the usual "cw issue", then it should be posted in the discourse section, not in the translation. So is it rather "issues encountered by the translator so that his translation is not fully ready?"

If the latter, I guess a smaal rewording would be good to avoid confusion.

Author: It's about issues found during **review**, so on a pending translation. There's no **issue** label there :(

Reviewer 4: oh.... x)

So maybe here again, we could talk about this more from the translator point of view: 

> if issues with the translation are found during its review, they will be discussed inside the comment section of the translation so that the translator can be notified and handle them.

or somethigng like that?

Author: ```suggestion
When issues with the translation are found during its review, they should be reported and discussed in the comment section of the translation so that the translator can be notified and handle them.
```

Feel free to edit this suggestion if you have better idea.

Reviewer 4: perfect :+1:

> - Unfortunately, approval of translation cannot be easily blocked by raising issues. If translation has many issues, cannot be easily fixed, and there's justified concern that it could be approved prematurely while not keeping up to quality standards, **it should be rejected**.

Reviewer 3: ```suggestion
- Unfortunately, the approval of a translation cannot be easily blocked by raising issues. If a translation has many issues, cannot be easily fixed, and there's justified concern that it could be approved prematurely while not keeping up to quality standards, **it should be rejected**.
```

> - Unfortunately, the approval of a translation cannot be easily blocked by raising issues. If a translation has many issues, cannot be easily fixed, and there's justified concern that it could be approved prematurely while not keeping up to quality standards, **it should be rejected**.

Reviewer 4: ```suggestion
- If a translation has too many issues or cannot be easily fixed, and there's justified concern that it could be approved prematurely while not keeping up to quality standards, **it can be rejected**.
```

no need to emphasises the hole in the process x) => just state what can happen/be done.

note: "should" -> "can" because the current guidelines is fro the translator more than the reviewer

Author: I hoped them to be useful for both: aurhors while creating, and reviewers for verification. But yeah, wording usually suggests author.

Reviewer 4: yes sure. But iirc, there is a "dedicated doc to be" about translations approval. That's why I suggests this here.

Author: The doc you mean somehow shifted and drifted and split into two parts. Quality guidelines (the ones you are reading) are meant to be applicable for authors, *and* for reviewers, as a kind of checklist they can use before approving new content.

I will review wording of all existing guidelines and try to balance it so they do not suggest authors as much, and take approves more into picture.

I was also thinking about another document, kind of manual, which would describe authoring and review process from functional side: views, editors, actions, options, etc. I have no good idea for this yet.

> - After translation is approved, it becomes **responsibility of all the parties: the approver, the translator, and the kata author** (in this order) to fix any potential issues which might come up later, when users attempt to submit their solutions.

Reviewer 3: ```suggestion
- After a translation is approved, it becomes **the responsibility of all parties involved: the user who approved the translation, the translator, and the kata author** (in this order) to fix any potential issues which might come up later when users attempt to submit their solutions.
```

> - After a translation is approved, it becomes **the responsibility of all parties involved: the user who approved the translation, the translator, and the kata author** (in this order) to fix any potential issues which might come up later when users attempt to submit their solutions.

Reviewer 4: I disagree on the order, in this one.
I understand the reason (showing the great responsability of the approver), but the first responsible should imo be the translator himself. If the pressure is on the guy who didn't write the code, that doesn't sound reasonnable to me. And if it's officially documented that way:
* some morons (translators) will hide behind that
* if the kata author trusted the translator while he doesn't know the language, he cannot actually do anything so the order is biased
* if it actually turns that way and the approver actually has to handle the mess, very few translations will be aproved in the near future...

Author: Sounds very reasonable, but I wanted to make approves aware that they are not just pushing some green buttons for fun.

Reviewer 4: Yup, I got that. It's effectively hard to state this clearly.
Maybe it's just not the appropriate place? Like, keeping here focus on the translator, and in the "approving/reviewing translations" doc, make strong evidence about the responsibility of the approver/reviewer?

Author: I think I will change the order as you suggest, and remove the strong note about the order.

Author: ```suggestion
- After a translation is approved, it becomes **the responsibility of all parties involved: the translator, the user who approved the translation, and the kata author**, to fix any potential issues which might come up later when users attempt to submit their solutions.
```

Reviewer 4: :+1:

This article contains a set of guidelines, which can be used by translators to create good translations to existing kata. They were collected to help ensure that translations are of sufficient quality and users' experience will be as good as possible.

The guidelines should be used by translators and reviewers to verify whether translations about to be introduced into Codewars are of sufficient quality. Conformance with them should be verified before the translation is published and before it's approved.

> Failing to comply with the below guidelines should be reported as an issue or a comment requesting for a fix. In case of severe violations, affected translation can be rejected, or even deleted.

Reviewer 3: ```suggestion
Failure to comply with the below guidelines is considered an issue and should be reported as such. In case of severe violations, the affected translation can be rejected or even deleted.
```

## General translation guidelines

> - **Respect [General Content Guidelines](/recipes/authoring/general/).**

Reviewer 3: ```suggestion
- **Respect the [General Content Guidelines](/recipes/authoring/general/).**
```

> - **Translations should improve the overall quality of the kata.** Sub-optimal quality of existing, approved translations should not be used as an argument for low quality of the translation. In particular:

Reviewer 3: ```suggestion
- **Translations should improve the overall quality of the kata.** The sub-optimal quality of existing, approved translations should not be used as an excuse for authoring a low-quality translation. In particular:
```

>   - Description can be clarified, if necessary.

Reviewer 3: ```suggestion
- The description can be clarified if necessary.
```

>   - Test suite can use different structure than other translations. **[Random tests](/recipes/authoring/kata-snippets/full-tests/#random-tests), [full test suite](/recipes/authoring/kata-snippets/full-tests/) and  [sample tests](/recipes/authoring/kata-snippets/sample-tests/) should conform to Codewars quality guidelines.**

Reviewer 3: ```suggestion
- The test suite may employ a different structure from that of other translations. **The [sample](/recipes/authoring/kata-snippets/sample-tests/) and [submit](/recipes/authoring/kata-snippets/full-tests/) tests (including [random tests](/recipes/authoring/kata-snippets/full-tests/#random-tests)) should conform to Codewars quality guidelines.**
```

> - If, while translation is created, some serious problems are spotted in other language versions, they should be fixed or raised as appropriate issues in the kata discourse.

Reviewer 3: ```suggestion
- If serious issues with other language versions are uncovered in the course of authoring a translation, they should be fixed or raised as appropriate issues in the kata discourse.
```

- **Translations should improve the overall quality of the kata.** The sub-optimal quality of existing, approved translations should not be used as an excuse for authoring a low-quality translation. In particular:
  - The description can be clarified if necessary.
  - The test suite may employ a different structure from that of other translations. **The [sample](/recipes/authoring/kata-snippets/sample-tests/) and [submission](/recipes/authoring/kata-snippets/full-tests/) tests (including [random tests](/recipes/authoring/kata-snippets/full-tests/#random-tests)) should conform to Codewars quality guidelines.**
  - If serious issues with other language versions are uncovered while translating, they should be fixed in the translation and appropriate issues should be raised about all other concerned languages in the kata discourse.

## Conformity to overall kata idea

> A translation should not substantially change either the overall idea nor the requirements of the kata. It should be as consistent as possible with the original version in terms of difficulty, required performance, and overall composition. In particular:

Reviewer 3: ```suggestion
A translation should substantially change neither the overall idea nor the requirements of the kata. It should be as consistent as possible with the original version in terms of difficulty, required performance, and overall composition. In particular:
```

## Conformance to overall kata idea

A translation should not substantially change the overall idea of the kata. It should be as consistent as possible with the original version in terms of difficulty, required performance, and overall composition. In particular:

> - **Respect the performance requirements of the kata.** If, for example, it's agreed that kata should accept only (sub)linear solutions, make sure that translated tests reject solutions of higher computational complexity. Languages and runtimes differ in terms of performance, so increase/decrease amount of test cases and input ranges when necessary.

Reviewer 3: ```suggestion
- **Respect the performance requirements of the kata.** If, for example, it's agreed that the kata should accept only (sub)linear solutions, make sure that the translated tests reject solutions of higher computational complexity. Languages and runtimes differ in terms of performance, so increase/decrease the amount of test cases and input ranges accordingly.
```

## Code components

- **Respect the [General Coding Guidelines](/recipes/authoring/kata-snippets/coding-general/), any conventions imposed by your [language](/languages/) and guidelines specific to each component of the kata**, e.g. the [preloaded](/recipes/authoring/kata-snippets/preloaded/) section, [reference solution](/recipes/authoring/kata-snippets/reference-solution/), [sample tests](/recipes/authoring/kata-snippets/sample-tests/) and [submission tests](/recipes/authoring/kata-snippets/full-tests/).
- If possible, **sample tests and fixed tests should remain consistent between translations.** It may not always be feasible due to inherent differences in programming languages, but doing so can save follow-up questions on the language a user is attempting the kata in when a question on failing tests is raised in the kata discourse.

> - Sometimes, the description is not the best place to put some language specific information. For example, the translator may wish to add some information on arguments, data formats, memory management, etc. and keep the description language-agnostic at the same time. In such cases, it's allowed to put such information as an explanatory comments in the solution setup.

Reviewer 3: ```suggestion
- Sometimes, the description is not the best place to put some language specific information. For example, the translator may wish to add some information on arguments, data formats, memory management, etc. and keep the description language-agnostic at the same time. In such cases, it's allowed to put such information as an explanatory comment in the solution setup.
```

## Code components

> - **Respect [General Coding Guidelines](/recipes/authoring/kata-snippets/coding-general/), coding guidelines of your [language](/languages/), and guidelines specific for particular kata code components** ([sample tests](/recipes/authoring/kata-snippets/sample-tests/), [full test suite](/recipes/authoring/kata-snippets/full-tests/), [preloaded code](/recipes/authoring/kata-snippets/preloaded/) and [proposed solution](/recipes/authoring/kata-snippets/proposed-solution/)).

Reviewer 3: ```suggestion
- **Respect the [General Coding Guidelines](/recipes/authoring/kata-snippets/coding-general/), any conventions imposed by your [language](/languages/) and guidelines specific to each component of the kata**, e.g. the [preloaded](/recipes/authoring/kata-snippets/preloaded/) section, [complete solution](/recipes/authoring/kata-snippets/proposed-solution/), [sample tests](/recipes/authoring/kata-snippets/sample-tests/) and [submit tests](/recipes/authoring/kata-snippets/full-tests/).
```

> - If possible, **sample tests and fixed tests should use the same inputs in all translations.** It's not always possible between languages, but when done, it makes support of users much easier. When a question about failed test pops up in kata discourse, it can save follow-up questions for attempted language, etc.

Reviewer 3: ```suggestion
- If possible, **sample tests and fixed tests should remain consistent between translations.** It may not always be feasible due to inherent differences in programming languages, but doing so can save follow-up questions on the language a user is attempting the kata in when a question on failing tests is raised in the kata discourse.
```

Author: English is difficult.
